lightning/src/ln/peers/handshake/mod.rs

```diff
@@ -0,0 +1,398 @@
+			ephemeral_private_key,
+			remote_public_key,
+			chaining_key,
+			&mut hash,
+		);
+
+		self.state = Some(HandshakeState::AwaitingActTwo(ActTwoExpectation {
+			hash,
+			chaining_key,
+			temporary_key,
+			ephemeral_private_key: (*ephemeral_private_key).clone(),
+		}));
+
+		Ok(ActOne(act_one))
+	}
+
+	pub(crate) fn process_act_one(&mut self, act: ActOne, ephemeral_private_key: &SecretKey) -> Result<ActTwo, String> {
+		let state = self.state.take();
+		let act_one_expectation = match state {
+			Some(HandshakeState::AwaitingActOne(act_state)) => act_state,
+			Some(HandshakeState::Blank) => {
+				// this can also be initiated from a blank state
+				// public key
+				let public_key = Self::private_key_to_public_key(&self.private_key);
+				let (hash, chaining_key) = Self::initialize_state(&public_key);
+				ActOneExpectation {
+					hash,
+					chaining_key,
+				}
+			}
+			_ => {
+				self.state = state;
+				return Err("unexpected state".to_string());
+			}
+		};
+
+		let mut hash = act_one_expectation.hash;
+		let (remote_ephemeral_public_key, chaining_key, _) = self.process_act_message(
+			act.0,
+			&self.private_key,
+			act_one_expectation.chaining_key,
+			&mut hash,
+		)?;
+
+		let (act_two, chaining_key, temporary_key) = self.calculate_act_message(
+			ephemeral_private_key,
+			&remote_ephemeral_public_key,
+			chaining_key,
+			&mut hash,
+		);
+
+		self.state = Some(HandshakeState::AwaitingActThree(ActThreeExpectation {
+			hash,
+			chaining_key,
+			temporary_key,
+			ephemeral_private_key: (*ephemeral_private_key).clone(),
+			remote_ephemeral_public_key,
+		}));
+
+		Ok(ActTwo(act_two))
+	}
+
+	pub(crate) fn process_act_two(&mut self, act: ActTwo) -> Result<(ActThree, Conduit), String> {
+		let state = self.state.take();
+		let act_two_expectation = match state {
+			Some(HandshakeState::AwaitingActTwo(act_state)) => act_state,
+			_ => {
+				self.state = state;
+				return Err("unexpected state".to_string());
+			}
+		};
+
+		let mut hash = act_two_expectation.hash;
+		let (remote_ephemeral_public_key, chaining_key, temporary_key) = self.process_act_message(
+			act.0,
+			&act_two_expectation.ephemeral_private_key,
+			act_two_expectation.chaining_key,
+			&mut hash,
+		)?;
+
+		self.state = Some(HandshakeState::Complete);
+
+		// start serializing act three
+
+		let static_public_key = Self::private_key_to_public_key(&self.private_key);
+		let tagged_encrypted_pubkey = chacha::encrypt(&temporary_key, 1, &hash.value, &static_public_key.serialize());
+		hash.update(&tagged_encrypted_pubkey);
+
+		let ecdh = Self::ecdh(&self.private_key, &remote_ephemeral_public_key);
+		let (chaining_key, temporary_key) = hkdf::derive(&chaining_key, &ecdh);
+		let authentication_tag = chacha::encrypt(&temporary_key, 0, &hash.value, &[0; 0]);
+		let (sending_key, receiving_key) = hkdf::derive(&chaining_key, &[0; 0]);
+
+		let mut act_three_vec = [0u8].to_vec();
```

Comment:
We know how long this is going to be, we should just allocate a buffer instead of malloc()ing (and a few other places in this file).